src/rotation.zig

```diff
@@ -321,6 +321,193 @@ pub const RotationModes = struct {
+			enabledConnections: u6,
+
+			pub fn isConnected(self: @This(), neighbor: Neighbor) bool {
+				return (self.enabledConnections & Neighbor.bitMask(neighbor)) != 0;
+			}
+
+			pub fn setConnection(self: *@This(), neighbor: Neighbor, value: bool) void {
+				if (value) {
+					self.enabledConnections |= Neighbor.bitMask(neighbor);
+				} else {
+					self.enabledConnections &= ~Neighbor.bitMask(neighbor);
+				}
+			}
+		};
+
+		fn init() void {
+			branchModels = .init(main.globalAllocator.allocator);
+		}
+
+		fn deinit() void {
+			branchModels.deinit();
+		}
+
+		fn branchTransform(quad: *main.models.QuadInfo, data: BranchData) void {
+			for(&quad.corners) |*corner| {
+				if(
+					(!data.isConnected(Neighbor.dirNegX) and corner[0] == 0)
+					or (!data.isConnected(Neighbor.dirPosX) and corner[0] == 1)
+					or (!data.isConnected(Neighbor.dirNegY) and corner[1] == 0)
+					or (!data.isConnected(Neighbor.dirPosY) and corner[1] == 1)
+					or (!data.isConnected(Neighbor.dirDown) and corner[2] == 0)
+					or (!data.isConnected(Neighbor.dirUp) and corner[2] == 1)
+				) return degenerateQuad(quad);
+			}
+		}
+
+		fn degenerateQuad(quad: *main.models.QuadInfo) void {
+			for(&quad.corners) |*corner| {
+				corner[0] = 0.5;
+				corner[1] = 0.5;
+				corner[2] = 0.5;
+			}
+		}
+
+		pub fn createBlockModel(modelId: []const u8) u16 {
+			if(branchModels.get(modelId)) |modelIndex| return modelIndex;
+
+			const baseModelIndex = main.models.getModelIndex(modelId);
+			const baseModel = main.models.models.items[baseModelIndex];
+
+			const modelIndex: u16 = baseModel.transformModel(branchTransform, .{@as(BranchData, @bitCast(@as(u6, 0)))});
+			for(1..64) |branchData| {
+
+				_ = baseModel.transformModel(branchTransform, .{@as(BranchData, @bitCast(@as(u6, @intCast(branchData))))});
+			}
+			branchModels.put(modelId, modelIndex) catch unreachable;
+			return modelIndex;
+		}
+
+		pub fn model(block: Block) u16 {
+			return blocks.meshes.modelIndexStart(block) + (block.data & 63);
+		}
+
+		pub fn generateData(
+			_: *main.game.World,
+			_: Vec3i,
+			_: Vec3f,
+			_: Vec3f,
+			_: Vec3i,
+			neighbor: ?Neighbor,
+			currentBlock: *Block,
+			neighborBlock: Block,
+			blockPlacing: bool,
+		) bool {
+			const blockBaseModel = blocks.meshes.modelIndexStart(currentBlock.*);
+			const neighborBaseModel = blocks.meshes.modelIndexStart(neighborBlock);
+
+			if(
+				blockPlacing
+				or (blockBaseModel == neighborBaseModel)
+				or neighborBlock.solid()
+			) {
+				const neighborModel = blocks.meshes.model(neighborBlock);
+				const targetVal = (neighborBlock.solid() and  ((blockBaseModel == neighborBaseModel) or main.models.models.items[neighborModel].isNeighborOccluded[neighbor.?.reverse().toInt()]));
+
+				var currentData: BranchData = @bitCast(@as(u6, @truncate(currentBlock.data)));
+				// Branch block upon placement should extend towards a block it was placed
+				// on if the block is solid or also uses branch model.
```

Comment:
I think the comment would be more helpful, if the `targetVal` declaration was also below it.